scipy/optimize/_linprog_highs.py

```diff
@@ -403,6 +402,10 @@ def _linprog_highs(lp, solver, time_limit=None, presolve=True,
     status, message = statuses[res['status']]
 
     x = np.array(res['x']) if 'x' in res else None
+    if np.any(x) and integrality is not None:
```

Comment:
I'll have to do some digging to see if there is any rounding going on upstream when MIP solutions are returned

Reply:
It doesn't look like it. Without this line, `test_mip1` (the `n=4` magic square problem) fails.

Reply:
I'll create an upstream issue asking about this

Reply:
For that problem I find the answers end up being really close to integral (usually within a tolerance of 1e-15).  When HiGHS prints out a solution file, they round using a tolerance of [`1e-13`](https://github.com/ERGO-Code/HiGHS/blob/8e0f2a523281fe6c6eca29236fcdb89552d1ed7c/src/lp_data/HighsModelUtils.cpp#L213).  I'm not sure what is better: upstream or downstream being responsible for rounding.  I guess if you asked a MIP solver for a variable that integral, it should probably return an integral value (not something really close to integral)?

I'll open an upstream issue to at least ask the question (I'm probably fine with whatever the answer turns out being): https://github.com/ERGO-Code/HiGHS/issues/649

Reply:
> I guess if you asked a MIP solver for a variable that integral, it should probably return an integral value (not something really close to integral)?

Maybe I'm wrong, but I think so. Perhaps I'm forgetting, but I seem to remember it being natural for LP algorithms to ensure that bound constraints on decision variables are strictly satisfied. My reasoning would be that if bound constraints are exactly satisfied, then integral constraints on DVs should be exactly satisfied. Of course, there will always be some error in the satisfaction of equality constraints - but that is where the error should be, IMO, not in the values of the DVs.